artemis-server/src/main/java/org/apache/activemq/artemis/core/management/impl/QueueControlImpl.java

```diff
@@ -883,17 +883,12 @@ protected Map<String, Object>[] getFirstMessage() throws Exception {
       clearIO();
       try {
          List<Map<String, Object>> messages = new ArrayList<>();
-         queue.flushExecutor();
          final int attributeSizeLimit = addressSettingsRepository.getMatch(address).getManagementMessageAttributeSizeLimit();
-         try (LinkedListIterator<MessageReference> iterator = queue.browserIterator()) {
-            // returns just the first, as it's the first only
-            if (iterator.hasNext()) {
-               MessageReference ref = iterator.next();
-               Message message = ref.getMessage();
-               messages.add(message.toMap(attributeSizeLimit));
-            }
-            return messages.toArray(new Map[1]);
+         MessageReference firstMessage = queue.peekFirstMessage();
+         if (firstMessage != null) {
+            messages.add(firstMessage.getMessage().toMap(attributeSizeLimit));
          }
+         return messages.toArray(new Map[1]);
```

Comment:
If the idea is to optimise this overall getFirstMessage() method, then there still looks to be room for improvement here. It creates an ArrayList of Map, then afterwards figures out whether it will put 0 or 1 things in it, causing creation of a new array of capacity 10 if there is, and then converts the resulting ArrayList of 0/1 items into a new created array of size 1.

Seems like it could just omit the ArrayList and create the final array, then if the map value exists add it, then return it. Or if allowed by the API and its uses, even just only create the array once its known there is something to put in it, and return null otherwise.

Reply:
it makes sense... after reading what it does.. 

I was just getting rid of the iterator() which caused an outage on an user I was helping out.

But I will improve that part as well.. thanks